src/main/scala/akka/persistence/cassandra/journal/StreamMergerActor.scala

```diff
@@ -0,0 +1,134 @@
+package akka.persistence.cassandra.journal
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+import scala.util.control.Exception._
+
+import akka.actor.{Props, Actor}
+import akka.persistence.cassandra.journal.StreamMerger._
+import akka.serialization.SerializationExtension
+import com.datastax.driver.core.Session
+import com.datastax.driver.core.exceptions.DriverException
+
+object StreamMergerActor {
+  def props(config: CassandraJournalConfig, session: Session): Props =
+    Props(new StreamMergerActor(config, session))
+
+  val name = "StreamMerger"
+}
+
+/**
+ * Merges n independent physical streams into a single logical stream.
+ * Preserves causal order per persistenceId.
+ *
+ * The intent of current functionality is to work in batches. The merger retrieves data from all
+ * streams up to a maximum of configured step. Then creates a single logical stream out of the
+ * n independent streams. There is no ordering between the independent streams and the only
+ * requirement is to maintain causal order per persistenceId. When the whole batch can be merged
+ * the stream is emitted. If not (e.g. causality violation in given batch) then the merger
+ * requests again for more data.
+ *
+ * The emmited logical stream can be stored into database in a batch. This grants atomicity, but not
+ * isolation which should be what we need. We can recover with RYOW consistency from the persisted
+ * index table similarly to original functionality. But if we store the progress in all independent
+ * streams we should be able to rerun the merging and idempotency shoud ensure we don't emit
+ * duplicates. If part of the batch is not perceived as complete due to eventual
+ * consistency should be fine and reconciled during replay.
+ *
+ * Other options were considered, e.g. storing each element as it arrives. This would require a more
+ * granular control of processed elements. E.g. updating the master table with a "processed" flag
+ * or storing current state of a buffer in database using a batch. The reason is that maintaining
+ * causal order requires a buffer to store out of sequence elements until they are expected in
+ * sequence. This buffer must be persistent and recovered in case of failure otherwise the elements
+ * could remain unprocessed. Creating a batch does not require any additional data structure and
+ * is similar to how CassandraJournal works.
```

Comment:
@krasserm please review if this logic is valid or if you can think of a better solution?